text/0000-feature-pruning.md

```diff
@@ -0,0 +1,224 @@
+usage might be
+
+```
+{
+  "dependencies": {
+    "chalk": "*"
+  },
+  "devDependencies": {
+    "ava": "*",
+    "webpack": "*",
+    "selenium-webdriver": "*"
+  },
+  "features": {
+    "build": ["webpack"],
+    "unit-test": ["ava"],
+    "integ-test": ["selenium-webdriver"],
+    "test": ["unit-test", "integ-test"]
+  }
+}
+```
+
+The features hash will contain an arbitrary number of user-defined keys which
+correspond to flags passed to `yarn prune`.  The values associated with those
+keys are always lists, consisting of either objects or strings.  If passed a
+string, it must be one of the following: a key in the `dependencies`,
+`devDependencies`, `peerDependencies` or `optionalDependencies`; the name of
+another key in the `features` hash; or one of `dependencies`,
+`devDependencies`, `peerDependencies` or `optionalDependencies`.  If the
+name of a dependency is provided, that dependency is a member of the feature; if
+the name of a feature is provided, all dependencies in that feature are included
+in the parent feature; if one of `dependencies`, `devDependencies`,
+`peerDependencies` or `optionalDependencies`, all dependencies listed in the
+corresponding hash of package.json are considered members of the feature.  This,
+however, can lead to ambiguity; for instance if you want to have a feature be
+the name of one of your dependencies (for instance if you have a dependency that
+has peerDependencies and has a dedicated CI stage, like a a11y test).  In this
+case, an object may be passed for disambiguation.  The object must contain two
+keys, a `type`, which is one of { `feature`, `dependency`, `package.json` }, and
+a `name`, which is the name of the feature or dependency, or the key to a
+package.json field.  Thus, the following package.json is equivalent to the
+example above
+
+```
+{
+  "dependencies": {
+    "chalk": "*"
+  },
+  "devDependencies": {
+    "ava": "*",
+    "webpack": "*",
+    "selenium-webdriver": "*"
+  },
+  "features": {
+    "build": [{"type": "dependency", "name": "webpack"}],
+    "unit-test": [{"type": "dependency", "name": "ava"}],
+    "integ-test": [{"type": "dependency", "name": "selenium-webdriver"}],
+    "test": [{ "type": "feature", "name": "unit-test"},
+             { "type": "feature", "name": "integ-test"}]
+  }
+}
+```
+
+## Yarn prune
+
+I also propose adding a new cli command, `yarn prune`, named for the equivalent
+[npm command](https://docs.npmjs.com/cli/prune), which it approximates.  For npm
+compatibility, a `yarn prune`, with no arguments, would remove any dependencies
+that are not needed.  In this way, it is similar to a `yarn install --force`,
+except that it is strictly subtractive, rather than additive.  
+
+However, unlike the existing npm command, when flags are passed to `yarn prune`,
+the set of dependencies specified by the feature whose name corresponds to the
+flag will be retained, as well as all transitive dependencies, while the rest of

```

Comment:
I understand why you propose it like this, but it seems weird that the passed argument is retained, _not_ pruned. I.e. intuitively I'd think that `yarn prune --build` will prune the "build" feature, not retain it.